runtime/src/iree-amd-aie/driver/xrt-lite/direct_command_buffer.cc

```diff
@@ -0,0 +1,216 @@
+  void* target_device_buffer_ptr = target_device_buffer->map();
+  iree_device_size_t target_offset =
+      iree_hal_buffer_byte_offset(target_ref.buffer) + target_ref.offset;
+
+  shim_xdna::bo* source_device_buffer = iree_hal_xrt_lite_buffer_handle(
+      iree_hal_buffer_allocated_buffer(source_ref.buffer));
+  void* source_device_buffer_ptr = source_device_buffer->map();
+  iree_device_size_t source_offset =
+      iree_hal_buffer_byte_offset(source_ref.buffer) + source_ref.offset;
+
+  uint8_t* dst = (uint8_t*)target_device_buffer_ptr + target_offset;
+  uint8_t* src = (uint8_t*)source_device_buffer_ptr + source_offset;
+  memcpy(dst, src, target_ref.length);
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_xrt_lite_direct_command_buffer_dispatch(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_executable_t* base_executable, int32_t entry_point,
+    const uint32_t workgroup_count[3], iree_const_byte_span_t constants,
+    iree_hal_buffer_ref_list_t bindings, iree_hal_dispatch_flags_t flags) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_direct_command_buffer* command_buffer =
+      reinterpret_cast<iree_hal_xrt_lite_direct_command_buffer*>(
+          base_command_buffer);
+  // Lookup kernel parameters used for side-channeling additional launch
+  // information from the compiler.
+  iree_hal_xrt_lite_native_executable* executable =
+      reinterpret_cast<iree_hal_xrt_lite_native_executable*>(base_executable);
+  iree_hal_xrt_lite_kernel_params kernel_params =
+      executable->entry_points[entry_point];
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                       &executable));
+
+  size_t ctrl_code_size = kernel_params.asm_inst.size() * sizeof(uint32_t);
+  auto bo_ctrl_code = command_buffer->shim_device->alloc_bo(
+      ctrl_code_size, XCL_BO_FLAGS_CACHEABLE);
+  uint32_t* instr_buffer = static_cast<uint32_t*>(bo_ctrl_code->map());
+  memcpy(instr_buffer, kernel_params.asm_inst.data(), ctrl_code_size);
+  bo_ctrl_code->sync(shim_xdna::direction::host2device);
+
+  shim_xdna::kernel ebuf(command_buffer->shim_device->get_pdev(), ERT_START_CU);
+  std::unique_ptr<shim_xdna::hw_ctx> context =
+      command_buffer->shim_device->create_hw_context(kernel_params.pdi,
+                                                     kernel_params.kernel_name);
+  shim_xdna::cuidx_t cu_idx =
+      context->open_cu_context(kernel_params.kernel_name);
+
+  ebuf.set_cu_idx(cu_idx);
+  unsigned int opcode = 3;
+  ebuf.add_arg_64(opcode);
+  ebuf.add_arg_bo(*bo_ctrl_code);
+  ebuf.add_arg_32(kernel_params.asm_inst.size());
+
+  for (iree_host_size_t j = 0; j < bindings.count; ++j) {
+    shim_xdna::bo* bo = iree_hal_xrt_lite_buffer_handle(
+        iree_hal_buffer_allocated_buffer(bindings.values[j].buffer));
+    ebuf.add_arg_bo(*bo);
+    bo->sync(shim_xdna::direction::host2device);
```

Comment:
I think this should not be needed as we are doing this when we map/unmap the buffers, but I know the xrt hal is doing the same thing, so maybe add a TODO about this?

Reply:
Same